Create benchmark suite

nowells opened the pull request:
Extracted from https://github.com/siimon/prom-client/pull/220

KevinAMurray commented on benchmarks/registry.js:
```diff
@@ -0,0 +1,37 @@
+	add('metrics', () => {
+		registry.metrics();
+	});
+}
```
Is it worth adding something like:

        add('histogramObserve', () => {
                histogram.observe(1);
        });

(I've tinkered with performance changes that hit the observe() call in exchange for scrape benefits.  This would trigger a fail, but given the consequences of that I think a fail on observe() performance is good as it makes for a debate on whether or not that is acceptable.)

SimenB replied:
I think adding a suite for updating a metric, and not just serializing them, is very much valuable!

nowells replied:
Agreed! I'll add that to this PR. Currently I am working on adding logic to allow acceptable margin of error to speed differences (to account fro no-op performance PRs and run variance)

zbjornson commented on package.json:
```diff
@@ -31,15 +33,20 @@
 		"husky": "^0.14.3",
 		"jest": "^23.6.0",
 		"lint-staged": "^7.0.0",
+		"lodash": "^4.17.11",
```
Minor, and y'all can ignore this if you think it doesn't matter, but what do you think of depending only on `lodash/flatten`, and doing the rest with vanilla JS?

`forEach`, `map` and `keys` are easy with vanilla JS. `Object.values` is node 7+, so you'd have to use a `for...in` loop or similar.  `flatMap` could be done with `_.flatten(arr).map(...)`.

(❤️ fewer and smaller dependencies)

nowells replied:
👍

zbjornson commented on bin/check-versions.sh:
```diff
@@ -0,0 +1,12 @@
+	echo "You currently have prom-client@${INSTALLED_VERSION} in devDependencies, but you need to have prom-client@${LATEST_VERSION}.";
+	echo 'Please run `npm install --save-dev --force prom-client`';
+	exit 1
+fi
```
FreeBSD doesn't have bash by default, and Windows basically never does. In another package I help maintain, we switched our bash script to a node.js script to avoid compat issues. (See https://github.com/Automattic/node-canvas/pull/754.)

nowells replied:
:+1: I was thinking about potentially doing the install at benchmark runtime instead? What do you think? Would avoid the nasty `--force` you have to run to install yourself as a devDep, and would let us run technically against multiple versions (I would install into a tmp directory and require from there). Thoughts?

nowells replied:
But at the very least, 👍 to moving this check into nodejs.

nowells replied:
I extracted the benchmark runner, it now installs the requested version(s) on demand.

nowells replied:
I will need to fix up my newly extracted module to be `node@6` compatible, but I will do that next. Hopefully this looks like a good API for the package. I should have it ready and fully tested on node@6 on Monday.

SimenB replied:
there is https://www.npmjs.com/package/libnpm now, btw

nowells commented:
Let me know if there is anything about the new benchmark suite that you would like to be changed. Thank you all for the great feedback in improving this solution!

SimenB commented on package.json:
```diff
@@ -12,10 +12,11 @@
 		"lint": "eslint .",
 		"test-unit": "jest",
-		"compile-typescript": "tsc index.d.ts --noImplicitAny",
+		"compile-typescript": "tsc index.d.ts --noImplicitAny --skipLibCheck",
```
?

nowells replied:
It was complaining about Set

```
> tsc index.d.ts --noImplicitAny
node_modules/@types/node/index.d.ts:813:38 - error TS2304: Cannot find name 'Set'.
813         allowedNodeEnvironmentFlags: Set<string>;
                                         ~~~
```

https://travis-ci.org/siimon/prom-client/jobs/431769692

This appeared to fix it, but I honestly am not up to speed on typescript to know the incantations to get it to work properly. I am assuming that some dep that is now pulled in pulls in a new @types definition that `tsc` is choking on?

SimenB replied:
Try to add `--target es6` instead

nowells replied:
Awesome. Thanks for the guidance.

SimenB commented on package.json:
```diff
@@ -12,10 +12,11 @@
 	"scripts": {
-		"test": "npm run lint && npm run compile-typescript && npm run test-unit",
+		"benchmarks": "node ./benchmarks/index.js",
+		"test": "npm run lint && npm run compile-typescript && npm run test-unit && npm run benchmarks",
```
not sure I wanna run the benchmarks as part of `test`

nowells replied:
Happy to remove it. We do not fail the build, just output report. What would you prefer? Remove entirely? Move to `prepublishOnly`? Something else?

KevinAMurray commented:
It would be really nice if you could optionally benchmark with different numbers of labels/label values -- that's where I've seen some unexpected performance loss/degredation.

What I'm thinking of is having a mode where the histogram benchmark tests the following setups:

- 1 Label with 64 values
- 2 Labels each with 8 values
- 4 Labels with 4 values in first two labels and 2 values in last two labels
- 4 Labels with 2 values in first two labels and 4 values in last two labels
- 6 Labels each with 2 values

At an abstract level, all of the above 5 cases should have a roughly similar performance (okay, some decrease as you add labels). 

Thoughts?

SimenB commented:
I think we can all of them as suites?

KevinAMurray commented:
Yes, each would be separate suite.

KevinAMurray commented:
Well 3 suites; 1 registry/metrics, 1 registry/getMetricsAsJSON and 1 observe

nowells commented:
@KevinAMurray is that what you were thinking?

KevinAMurray commented:
Yes! Almost perfect. 

Can you replicate the changes in registry.js

I.e. so you get things like
registry ➭ metrics#1 with 64 ...
registry ➭ metrics#2 with 8 ....
...
and 
registry ➭ getMetricsAsJSON#1 with 64 ....
registry ➭ getMetricsAsJSON#2 with 8 ....
....

I'll leave it to others to decide if the extended tests should be optional -- main point is to make sure we don't hit performance with extra labels (or with less labels) with any changes.

nowells commented:
@KevinAMurray done

```
❯ node benchmarks/index.js

# Benchmark Regression

## Installing baseline packages

http fetch GET 200 https://registry.npmjs.org/prom-client 623ms
+ prom-client@11.1.3
added 3 packages from 3 contributors in 10.795s
## Starting benchmark suite

### Progress:

- registry ➭ getMetricsAsJSON#1 with 64 ➭ current x 188,770 ops/sec ±6.54% (78 runs sampled)
- registry ➭ getMetricsAsJSON#1 with 64 ➭ prom-client@latest x 182,510 ops/sec ±6.70% (75 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 8 ➭ current x 185,078 ops/sec ±6.28% (79 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 8 ➭ prom-client@latest x 190,074 ops/sec ±6.15% (77 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 ➭ current x 192,033 ops/sec ±5.57% (83 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 ➭ prom-client@latest x 197,083 ops/sec ±5.02% (81 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 ➭ current x 190,434 ops/sec ±5.80% (79 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 ➭ prom-client@latest x 193,411 ops/sec ±5.80% (81 runs sampled)
- registry ➭ getMetricsAsJSON#6 with 2 ➭ current x 179,065 ops/sec ±6.89% (73 runs sampled)
- registry ➭ getMetricsAsJSON#6 with 2 ➭ prom-client@latest x 188,530 ops/sec ±6.25% (82 runs sampled)
- registry ➭ metrics#1 with 64 ➭ current x 63,436 ops/sec ±6.34% (78 runs sampled)
- registry ➭ metrics#1 with 64 ➭ prom-client@latest x 63,148 ops/sec ±6.03% (85 runs sampled)
- registry ➭ metrics#2 with 8 ➭ current x 66,293 ops/sec ±4.69% (84 runs sampled)
- registry ➭ metrics#2 with 8 ➭ prom-client@latest x 63,092 ops/sec ±5.65% (81 runs sampled)
- registry ➭ metrics#2 with 4 and 2 with 2 ➭ current x 65,332 ops/sec ±5.34% (83 runs sampled)
- registry ➭ metrics#2 with 4 and 2 with 2 ➭ prom-client@latest x 66,005 ops/sec ±5.43% (84 runs sampled)
- registry ➭ metrics#2 with 2 and 2 with 4 ➭ current x 64,387 ops/sec ±7.25% (82 runs sampled)
- registry ➭ metrics#2 with 2 and 2 with 4 ➭ prom-client@latest x 64,578 ops/sec ±5.96% (84 runs sampled)
- registry ➭ metrics#6 with 2 ➭ current x 65,740 ops/sec ±5.66% (81 runs sampled)
- registry ➭ metrics#6 with 2 ➭ prom-client@latest x 64,821 ops/sec ±4.78% (84 runs sampled)
- histogram ➭ observe#1 with 64 ➭ current x 75,117 ops/sec ±3.53% (86 runs sampled)
- histogram ➭ observe#1 with 64 ➭ prom-client@latest x 71,216 ops/sec ±4.54% (84 runs sampled)
- histogram ➭ observe#2 with 8 ➭ current x 47,900 ops/sec ±5.40% (85 runs sampled)
- histogram ➭ observe#2 with 8 ➭ prom-client@latest x 47,157 ops/sec ±5.17% (84 runs sampled)
- histogram ➭ observe#2 with 4 and 2 with 2 ➭ current x 48,161 ops/sec ±5.16% (81 runs sampled)
- histogram ➭ observe#2 with 4 and 2 with 2 ➭ prom-client@latest x 47,452 ops/sec ±5.46% (86 runs sampled)
- histogram ➭ observe#2 with 2 and 2 with 4 ➭ current x 47,703 ops/sec ±4.79% (84 runs sampled)
- histogram ➭ observe#2 with 2 and 2 with 4 ➭ prom-client@latest x 46,325 ops/sec ±5.93% (81 runs sampled)
- histogram ➭ observe#6 with 2 ➭ current x 49,295 ops/sec ±4.48% (86 runs sampled)
- histogram ➭ observe#6 with 2 ➭ prom-client@latest x 46,965 ops/sec ±4.52% (84 runs sampled)

### Results:

┌────────────────────────────────────────┬────────────────────┬────────────────────┐
│ registry                               │ current            │ prom-client@latest │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#1 with 64             │ 188770.0304176406  │ 182509.811580655   │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#2 with 8              │ 185078.4692551817  │ 190073.78561996316 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#2 with 4 and 2 with 2 │ 192032.60709110482 │ 197083.45633913463 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#2 with 2 and 2 with 4 │ 190434.3068864219  │ 193410.77044994908 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#6 with 2              │ 179064.72776159606 │ 188529.5170065524  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#1 with 64                      │ 63435.73735528739  │ 63147.75416215507  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#2 with 8                       │ 66293.31396373227  │ 63091.621569732044 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#2 with 4 and 2 with 2          │ 65332.25042870494  │ 66004.98810840782  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#2 with 2 and 2 with 4          │ 64387.306761978405 │ 64578.30784346195  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#6 with 2                       │ 65740.02356453652  │ 64821.00138371078  │
└────────────────────────────────────────┴────────────────────┴────────────────────┘

┌───────────────────────────────┬────────────────────┬────────────────────┐
│ histogram                     │ current            │ prom-client@latest │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#1 with 64             │ 75117.47957995425  │ 71216.04931193707  │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#2 with 8              │ 47900.413719600896 │ 47156.56534408018  │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#2 with 4 and 2 with 2 │ 48161.41792755436  │ 47452.047280902414 │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#2 with 2 and 2 with 4 │ 47703.11912309259  │ 46324.75393982887  │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#6 with 2              │ 49295.20020689832  │ 46965.3745704656   │
└───────────────────────────────┴────────────────────┴────────────────────┘

### Summary:

✓ registry ➭ getMetricsAsJSON#1 with 64 is 3.316% faster.
⚠ registry ➭ getMetricsAsJSON#2 with 8 is 2.628% acceptably slower.
⚠ registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 is 2.563% acceptably slower.
⚠ registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 is 1.539% acceptably slower.
⚠ registry ➭ getMetricsAsJSON#6 with 2 is 5.020% acceptably slower.
✓ registry ➭ metrics#1 with 64 is 0.4540% faster.
✓ registry ➭ metrics#2 with 8 is 4.830% faster.
⚠ registry ➭ metrics#2 with 4 and 2 with 2 is 1.019% acceptably slower.
⚠ registry ➭ metrics#2 with 2 and 2 with 4 is 0.2958% acceptably slower.
✓ registry ➭ metrics#6 with 2 is 1.398% faster.
✓ histogram ➭ observe#1 with 64 is 5.194% faster.
✓ histogram ➭ observe#2 with 8 is 1.553% faster.
✓ histogram ➭ observe#2 with 4 and 2 with 2 is 1.473% faster.
✓ histogram ➭ observe#2 with 2 and 2 with 4 is 2.889% faster.
✓ histogram ➭ observe#6 with 2 is 4.726% faster.
```

nowells commented:
For comparison, here was `11.1.3` vs `11.1.1`

```
❯ node benchmarks/index.js

# Benchmark Regression

## Installing baseline packages

http fetch GET 200 https://registry.npmjs.org/prom-client 800ms
+ prom-client@11.1.3
added 3 packages from 3 contributors and audited 3 packages in 2.151s
found 0 vulnerabilities

+ prom-client@11.1.1
added 3 packages from 3 contributors and audited 3 packages in 0.407s
found 0 vulnerabilities

## Starting benchmark suite

### Progress:

- registry ➭ getMetricsAsJSON#1 with 64 ➭ prom-client@11.1.3 x 198,600 ops/sec ±5.79% (82 runs sampled)
- registry ➭ getMetricsAsJSON#1 with 64 ➭ prom-client@11.1.1 x 48,740 ops/sec ±4.64% (84 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 8 ➭ prom-client@11.1.3 x 201,914 ops/sec ±5.02% (82 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 8 ➭ prom-client@11.1.1 x 49,749 ops/sec ±4.86% (86 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 ➭ prom-client@11.1.3 x 196,092 ops/sec ±6.23% (80 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 ➭ prom-client@11.1.1 x 51,025 ops/sec ±4.10% (86 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 ➭ prom-client@11.1.3 x 199,202 ops/sec ±4.79% (84 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 ➭ prom-client@11.1.1 x 49,833 ops/sec ±4.50% (83 runs sampled)
- registry ➭ getMetricsAsJSON#6 with 2 ➭ prom-client@11.1.3 x 198,174 ops/sec ±5.52% (82 runs sampled)
- registry ➭ getMetricsAsJSON#6 with 2 ➭ prom-client@11.1.1 x 50,475 ops/sec ±4.24% (85 runs sampled)
- registry ➭ metrics#1 with 64 ➭ prom-client@11.1.3 x 69,446 ops/sec ±4.34% (85 runs sampled)
- registry ➭ metrics#1 with 64 ➭ prom-client@11.1.1 x 40,489 ops/sec ±4.73% (80 runs sampled)
- registry ➭ metrics#2 with 8 ➭ prom-client@11.1.3 x 68,442 ops/sec ±4.03% (83 runs sampled)
- registry ➭ metrics#2 with 8 ➭ prom-client@11.1.1 x 41,597 ops/sec ±4.16% (87 runs sampled)
- registry ➭ metrics#2 with 4 and 2 with 2 ➭ prom-client@11.1.3 x 69,123 ops/sec ±5.13% (87 runs sampled)
- registry ➭ metrics#2 with 4 and 2 with 2 ➭ prom-client@11.1.1 x 38,864 ops/sec ±6.12% (81 runs sampled)
- registry ➭ metrics#2 with 2 and 2 with 4 ➭ prom-client@11.1.3 x 67,610 ops/sec ±4.68% (85 runs sampled)
- registry ➭ metrics#2 with 2 and 2 with 4 ➭ prom-client@11.1.1 x 39,579 ops/sec ±5.51% (83 runs sampled)
- registry ➭ metrics#6 with 2 ➭ prom-client@11.1.3 x 68,565 ops/sec ±5.08% (84 runs sampled)
- registry ➭ metrics#6 with 2 ➭ prom-client@11.1.1 x 40,126 ops/sec ±5.45% (84 runs sampled)
- histogram ➭ observe#1 with 64 ➭ prom-client@11.1.3 x 73,381 ops/sec ±5.52% (82 runs sampled)
- histogram ➭ observe#1 with 64 ➭ prom-client@11.1.1 x 71,543 ops/sec ±5.94% (82 runs sampled)
- histogram ➭ observe#2 with 8 ➭ prom-client@11.1.3 x 47,900 ops/sec ±4.88% (85 runs sampled)
- histogram ➭ observe#2 with 8 ➭ prom-client@11.1.1 x 43,092 ops/sec ±5.97% (80 runs sampled)
- histogram ➭ observe#2 with 4 and 2 with 2 ➭ prom-client@11.1.3 x 46,920 ops/sec ±5.79% (82 runs sampled)
- histogram ➭ observe#2 with 4 and 2 with 2 ➭ prom-client@11.1.1 x 46,422 ops/sec ±5.19% (85 runs sampled)
- histogram ➭ observe#2 with 2 and 2 with 4 ➭ prom-client@11.1.3 x 47,604 ops/sec ±5.54% (82 runs sampled)
- histogram ➭ observe#2 with 2 and 2 with 4 ➭ prom-client@11.1.1 x 42,625 ops/sec ±7.50% (78 runs sampled)
- histogram ➭ observe#6 with 2 ➭ prom-client@11.1.3 x 46,186 ops/sec ±5.89% (80 runs sampled)
- histogram ➭ observe#6 with 2 ➭ prom-client@11.1.1 x 46,618 ops/sec ±5.90% (84 runs sampled)

### Results:

┌────────────────────────────────────────┬────────────────────┬────────────────────┐
│ registry                               │ prom-client@11.1.3 │ prom-client@11.1.1 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#1 with 64             │ 198600.38388104943 │ 48740.34424469919  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#2 with 8              │ 201914.4998599047  │ 49748.79883182819  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#2 with 4 and 2 with 2 │ 196091.99704080977 │ 51024.93532021316  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#2 with 2 and 2 with 4 │ 199202.39995463137 │ 49832.67403609232  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#6 with 2              │ 198174.42556418513 │ 50474.946989329095 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#1 with 64                      │ 69445.7762941244   │ 40488.84733371689  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#2 with 8                       │ 68442.45204817792  │ 41597.165344793786 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#2 with 4 and 2 with 2          │ 69122.59679372127  │ 38864.0256148741   │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#2 with 2 and 2 with 4          │ 67609.83546860874  │ 39579.11887281806  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#6 with 2                       │ 68564.7605864277   │ 40125.69952342952  │
└────────────────────────────────────────┴────────────────────┴────────────────────┘

┌───────────────────────────────┬────────────────────┬────────────────────┐
│ histogram                     │ prom-client@11.1.3 │ prom-client@11.1.1 │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#1 with 64             │ 73380.8330104104   │ 71542.54708540671  │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#2 with 8              │ 47899.75465917791  │ 43092.49517173999  │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#2 with 4 and 2 with 2 │ 46920.40292653754  │ 46422.46433670481  │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#2 with 2 and 2 with 4 │ 47604.45304779353  │ 42625.01362165591  │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#6 with 2              │ 46186.057206884434 │ 46618.28694546566  │
└───────────────────────────────┴────────────────────┴────────────────────┘

### Summary:

✓ registry ➭ getMetricsAsJSON#1 with 64 is 75.46% faster.
✓ registry ➭ getMetricsAsJSON#2 with 8 is 75.36% faster.
✓ registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 is 73.98% faster.
✓ registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 is 74.98% faster.
✓ registry ➭ getMetricsAsJSON#6 with 2 is 74.53% faster.
✓ registry ➭ metrics#1 with 64 is 41.70% faster.
✓ registry ➭ metrics#2 with 8 is 39.22% faster.
✓ registry ➭ metrics#2 with 4 and 2 with 2 is 43.78% faster.
✓ registry ➭ metrics#2 with 2 and 2 with 4 is 41.46% faster.
✓ registry ➭ metrics#6 with 2 is 41.48% faster.
✓ histogram ➭ observe#1 with 64 is 2.505% faster.
✓ histogram ➭ observe#2 with 8 is 10.04% faster.
✓ histogram ➭ observe#2 with 4 and 2 with 2 is 1.061% faster.
✓ histogram ➭ observe#2 with 2 and 2 with 4 is 10.46% faster.
⚠ histogram ➭ observe#6 with 2 is 0.9272% acceptably slower.
```

KevinAMurray commented:
Something slightly wierd going on here -- I'm not seeing the labels on the textual output from metrics, or in the JSON form, other than the buckets' own le label.

(e.g. if I do something as crude as a console.log(registry.metrics()) in the setup call....)

Can some check this? (As it may well be I've got issues with my  version)

nowells commented:
@KevinAMurray shoot, I called `histogram.observe(1, labels)` instead of `histogram.observe(labels, 1)`. My bad!

nowells commented:
Well that explains why there wasn't really any variance between the various label permutations. I'll push the fix and updated outputs.

KevinAMurray commented:
I should have spotted that as well!

nowells commented:
`11.1.3` compared to `11.1.1`

```
❯ node benchmarks/index.js                                                                                                                                                                                                           [40/1985]

# Benchmark Regression

## Installing baseline packages

http fetch GET 200 https://registry.npmjs.org/prom-client 292ms
+ prom-client@11.1.3
added 3 packages from 3 contributors and audited 3 packages in 1.58s
found 0 vulnerabilities

+ prom-client@11.1.1
added 3 packages from 3 contributors and audited 3 packages in 0.447s
found 0 vulnerabilities

## Starting benchmark suite

### Progress:

- registry ➭ getMetricsAsJSON#1 with 64 ➭ prom-client@11.1.3 x 3,160 ops/sec ±6.85% (76 runs sampled)
- registry ➭ getMetricsAsJSON#1 with 64 ➭ prom-client@11.1.1 x 463 ops/sec ±7.49% (71 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 8 ➭ prom-client@11.1.3 x 2,597 ops/sec ±5.33% (82 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 8 ➭ prom-client@11.1.1 x 424 ops/sec ±8.35% (72 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 ➭ prom-client@11.1.3 x 2,201 ops/sec ±6.14% (82 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 ➭ prom-client@11.1.1 x 307 ops/sec ±7.44% (69 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 ➭ prom-client@11.1.3 x 1,970 ops/sec ±7.50% (78 runs sampled)
- registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 ➭ prom-client@11.1.1 x 320 ops/sec ±7.33% (74 runs sampled)
- registry ➭ getMetricsAsJSON#6 with 2 ➭ prom-client@11.1.3 x 1,833 ops/sec ±7.79% (79 runs sampled)
- registry ➭ getMetricsAsJSON#6 with 2 ➭ prom-client@11.1.1 x 274 ops/sec ±7.90% (69 runs sampled)
- registry ➭ metrics#1 with 64 ➭ prom-client@11.1.3 x 904 ops/sec ±6.70% (80 runs sampled)
- registry ➭ metrics#1 with 64 ➭ prom-client@11.1.1 x 416 ops/sec ±6.78% (77 runs sampled)
- registry ➭ metrics#2 with 8 ➭ prom-client@11.1.3 x 789 ops/sec ±7.10% (76 runs sampled)
- registry ➭ metrics#2 with 8 ➭ prom-client@11.1.1 x 351 ops/sec ±6.48% (74 runs sampled)
- registry ➭ metrics#2 with 4 and 2 with 2 ➭ prom-client@11.1.3 x 699 ops/sec ±6.41% (80 runs sampled)
- registry ➭ metrics#2 with 4 and 2 with 2 ➭ prom-client@11.1.1 x 286 ops/sec ±6.81% (77 runs sampled)
- registry ➭ metrics#2 with 2 and 2 with 4 ➭ prom-client@11.1.3 x 693 ops/sec ±5.63% (81 runs sampled)
- registry ➭ metrics#2 with 2 and 2 with 4 ➭ prom-client@11.1.1 x 293 ops/sec ±4.94% (77 runs sampled)
- registry ➭ metrics#6 with 2 ➭ prom-client@11.1.3 x 619 ops/sec ±5.47% (81 runs sampled)
- registry ➭ metrics#6 with 2 ➭ prom-client@11.1.1 x 242 ops/sec ±5.25% (77 runs sampled)
- histogram ➭ observe#1 with 64 ➭ prom-client@11.1.3 x 35,952 ops/sec ±5.26% (87 runs sampled)
- histogram ➭ observe#1 with 64 ➭ prom-client@11.1.1 x 36,677 ops/sec ±3.72% (87 runs sampled)
- histogram ➭ observe#2 with 8 ➭ prom-client@11.1.3 x 10,467 ops/sec ±4.63% (85 runs sampled)
- histogram ➭ observe#2 with 8 ➭ prom-client@11.1.1 x 10,631 ops/sec ±4.30% (86 runs sampled)
- histogram ➭ observe#2 with 4 and 2 with 2 ➭ prom-client@11.1.3 x 8,597 ops/sec ±4.36% (82 runs sampled)
- histogram ➭ observe#2 with 4 and 2 with 2 ➭ prom-client@11.1.1 x 8,406 ops/sec ±5.25% (79 runs sampled)
- histogram ➭ observe#2 with 2 and 2 with 4 ➭ prom-client@11.1.3 x 8,596 ops/sec ±4.90% (85 runs sampled)
- histogram ➭ observe#2 with 2 and 2 with 4 ➭ prom-client@11.1.1 x 8,867 ops/sec ±4.89% (79 runs sampled)
- histogram ➭ observe#6 with 2 ➭ prom-client@11.1.3 x 7,681 ops/sec ±4.09% (88 runs sampled)
- histogram ➭ observe#6 with 2 ➭ prom-client@11.1.1 x 7,538 ops/sec ±4.00% (85 runs sampled)

### Results:

┌────────────────────────────────────────┬────────────────────┬────────────────────┐
│ registry                               │ prom-client@11.1.3 │ prom-client@11.1.1 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#1 with 64             │ 3159.8580396287584 │ 463.3412306606676  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#2 with 8              │ 2597.081858906496  │ 423.6981181313897  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#2 with 4 and 2 with 2 │ 2201.423195536035  │ 306.70427789852863 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#2 with 2 and 2 with 4 │ 1969.7301915880078 │ 320.13929466035256 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ getMetricsAsJSON#6 with 2              │ 1833.211557278113  │ 273.7964688417815  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#1 with 64                      │ 904.1769521843279  │ 415.6208289205233  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#2 with 8                       │ 789.3742899569859  │ 351.099953214439   │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#2 with 4 and 2 with 2          │ 699.0258120357616  │ 286.1557179977738  │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#2 with 2 and 2 with 4          │ 692.8816448192186  │ 293.00903603551274 │
├────────────────────────────────────────┼────────────────────┼────────────────────┤
│ metrics#6 with 2                       │ 618.8182656075049  │ 241.7387914473113  │
└────────────────────────────────────────┴────────────────────┴────────────────────┘

┌───────────────────────────────┬────────────────────┬────────────────────┐
│ histogram                     │ prom-client@11.1.3 │ prom-client@11.1.1 │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#1 with 64             │ 35951.648841787035 │ 36677.441331063834 │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#2 with 8              │ 10466.815661679873 │ 10630.691331798324 │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#2 with 4 and 2 with 2 │ 8596.97603614281   │ 8405.706977947284  │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#2 with 2 and 2 with 4 │ 8596.003572568006  │ 8867.4807994267    │
├───────────────────────────────┼────────────────────┼────────────────────┤
│ observe#6 with 2              │ 7680.717546491208  │ 7538.438788877932  │
└───────────────────────────────┴────────────────────┴────────────────────┘

### Summary:

✓ registry ➭ getMetricsAsJSON#1 with 64 is 85.34% faster.
✓ registry ➭ getMetricsAsJSON#2 with 8 is 83.69% faster.
✓ registry ➭ getMetricsAsJSON#2 with 4 and 2 with 2 is 86.07% faster.
✓ registry ➭ getMetricsAsJSON#2 with 2 and 2 with 4 is 83.75% faster.
✓ registry ➭ getMetricsAsJSON#6 with 2 is 85.06% faster.
✓ registry ➭ metrics#1 with 64 is 54.03% faster.
✓ registry ➭ metrics#2 with 8 is 55.52% faster.
✓ registry ➭ metrics#2 with 4 and 2 with 2 is 59.06% faster.
✓ registry ➭ metrics#2 with 2 and 2 with 4 is 57.71% faster.
✓ registry ➭ metrics#6 with 2 is 60.94% faster.
⚠ histogram ➭ observe#1 with 64 is 1.979% acceptably slower.
⚠ histogram ➭ observe#2 with 8 is 1.542% acceptably slower.
✓ histogram ➭ observe#2 with 4 and 2 with 2 is 2.225% faster.
⚠ histogram ➭ observe#2 with 2 and 2 with 4 is 3.061% acceptably slower.
✓ histogram ➭ observe#6 with 2 is 1.852% faster.
```

nowells commented:
@KevinAMurray great catch. I really should have caught that... sorry about that.

SimenB commented:
Wow, those are crazy improvements. Great job!

nowells commented:
> Wow, those are crazy improvements. Great job!

Thanks!

zbjornson commented:
Looks awesome to me! `npm run benchmarks` are still part of `test` in `package.json` though, which @SimenB and @siimon suggested keeping separate.

nowells commented:
Thanks everyone for your help in evolving and landing this! Thank you for the great package!
